### PR TITLE
Valid data masking

### LINKED ISSUE #1449: Allow to influence "damage" of results
Use case: visualization of partial results  with `kind='single'` and possibly custom shape, but the UDF knows what parts of it are valid. Examples are SSB/WDD or possibly iCoM.

### LINKED ISSUE #1473: Make `damage` available in `get_results()`
When calculating derived results in `UDF.get_results()` it can be useful to know the current `damage`, i.e. which parts of a buffer contain valid data. Offset and linear gradient correction in CoMUDF #1458 would be a concrete example where only valid portions of a result buffer should be included in the correction.

When looking at `UDFRunner`, the damage is an additional result buffer that is kept outside of the UDF. It is only merged with the UDF results when the final results object is assembled.

Possibly the way how damage is handled in live plots could be an inspiration? Here we already have an infrastructure that is similar to `get_results()`, but aware of damage. One could consider keeping track of damage for each result buffer? SSB and WDD would be examples of `kind='single'` derived result buffers that are, in a way, actually `kind='nav'`.

### PR DESCRIPTION
**Pull request by sk1p**
Makes the "valid nav mask" available in UDFs, and propagates it to results.

Closes #1473, closes #1449

## TODO

- [x] improve test cases
- [x] implement `BufferWrapper.masked_data` and friends
- [x] update plotting to use the per-buffer valid data mask instead of "global" damage
- [x] figure out how this fits best with `roi`s
- [x] add `raw_masked_data`
- [x] documentation

## Contributor Checklist:

* [x] I have added or updated my entry in [the creators.json file](https://github.com/LiberTEM/LiberTEM/blob/master/packaging/creators.json)
* [x] I have added [a changelog entry](https://github.com/LiberTEM/LiberTEM/tree/master/docs/source/changelog) for my contribution
* [x] I have added/updated documentation for all user-facing changes
* [x] I have added/updated test cases
* [ ] I have included the [rebuilt production build of the client](https://libertem.github.io/LiberTEM/contributing.html?#building-the-client) (only if changes were made to the GUI)

## Reviewer Checklist:

* [ ] `/azp run libertem.libertem-data` passed
* [ ] No import of GPL code from MIT code

### CODE REVIEW

**Comment by uellue** [comment]
I like the API! Really minimal and simple. Regarding masked array support in other backends, I came up with the following:

* Dask: https://docs.dask.org/en/latest/generated/dask.array.ma.masked_array.html
* CuPy: https://github.com/cupy/cupy/issues/2225
* Sparse: https://github.com/pydata/sparse/issues/359

Since we have, as far as I am aware, only NumPy and Dask results, it should work. If we want to support other return backends, one could implement the thin wrapper alternative for them, essentially a "poor man's masked array" type. When the time comes we can think of an API that allows creating the right kind of array which is backwards-compatible with NumPy masked arrays.

Regarding ROI behavior, what about the following?

Internally, everything works on the ROI like before. That means the mask is relative to the nav space with ROI applied -- always of the same shape as the actual result buffer one is interacting with. Result buffer `.data` only sets the values within the ROI that are not masked, meaning masked parts of result buffers are treated the same way as areas outside the ROI.

**Comment by uellue** [comment]
Hm, the `fill_value` parameter is interesting relative to our ROI considerations. One could also argue that masked areas should have the fill value and not the "outside of ROI" value.

Now I'm not so sure if we should use masked arrays directly.

## Pro

* We use an existing API of NumPy
* Potentially benefit from additional capabilities, see open question
* It generally matches what we want to express with the valid mask

## Con
* We should support all features of masked arrays in a predictable way if we allow users to pass them directly. With ROIs it leads to an ambiguous situation.
* The behavior of masked arrays is opposite of ROIs and damage -- "1" means invalid for masked arrays, but valid for ROI and damage.
* Not available in CuPy, which would be a likely target for `get_results()` implementations that benefit from GPUs, like iCoM

## Open questions

* Will the additional features of masked arrays be useful for the user, such as numerical operations on them?

## Conclusion

Looking at the implications, we might be better off to implement a thin wrapper type?

**Comment by matbryan52** [comment]
> ## Open questions
> * Will the additional features of masked arrays be useful for the user, such as numerical operations on them?
> 
> ## Conclusion
> Looking at the implications, we might be better off to implement a thin wrapper type?

My view on this part is that if we go with a thin wrapper, we can always offer a method on the wrapper which gives the user a true `np.ma` array with the most useful properties for them (valid only, roi only, valid + roi etc), and it gives us more flexibility in terms of fill values because we can handle that ourselves. We also don't have to construct anything new unless the user actually wants to take advantage of `np.ma`, which might avoid some copies ?

We also retain the option to support non-masked-compatible backends, as you've said already.

Overall I like the implementation and idea behind it 👍

**Comment by sk1p** [comment]
> The behavior of masked arrays is opposite of ROIs and damage -- "1" means invalid for masked arrays, but valid for ROI and damage.

Oof, yeah, I completely missed that one. I think that alone is enough to convince me that a thin wrapper can provide a better API.

> > Looking at the implications, we might be better off to implement a thin wrapper type?
> 
> My view on this part is that if we go with a thin wrapper, we can always offer a method on the wrapper which gives the user a true `np.ma` array with the most useful properties for them (valid only, roi only, valid + roi etc), and it gives us more flexibility in terms of fill values because we can handle that ourselves. We also don't have to construct anything new unless the user actually wants to take advantage of `np.ma`, which might avoid some copies ?
> 
> We also retain the option to support non-masked-compatible backends, as you've said already.

Thanks for the feedback, I'll work on an updated version of these changes. I'm thinking on what aspects have to be taken into account.

> Since we have, as far as I am aware, only NumPy and Dask results

Yeah, according to our documentation, `get_results` [is supposed to always return `np.ndarray`](https://libertem.github.io/LiberTEM/udf/advanced.html#post-processing-after-merging). This is actually wrong, as you correctly state, it can also be a `dask.array.core.Array`, with the delayed executor. I ran the test suite with the following patch, and at least for what we cover, `Union[np.ndarray, dask.array.core.Array]` is the true type there.

```patch
diff --git a/src/libertem/udf/base.py b/src/libertem/udf/base.py
index bc0b80b9..3efc03e3 100644
--- a/src/libertem/udf/base.py
+++ b/src/libertem/udf/base.py
@@ -1214,6 +1214,8 @@ class UDFBase(UDFProtocol):
         # wrap numpy results into `ResultBuffer`s:
         results = {}
         for name, arr in results_tmp.items():
+            import dask.array.core
+            assert isinstance(arr, (np.ndarray, dask.array.core.Array))
             self._check_results(decl, arr, name)
             buf_decl = decl[name]
             buf = results_buffer_cls[name](
```

We do not yet support returning `cupy` arrays from `get_results`, for the iCoM case, I'm currently doing a round-trip CPU->GPU->CPU, which is still very fast . In the future, if we want to optimize the "very large scan" use-case, we may want to keep the result buffers on the GPU, for example to directly apply a shader for visualization.

I'm also thinking about including some helper functions on the wrapper type, as there will be some common operations:

* Is everything valid?
* What is the (outer) bounding box (as 2d or nd slices)?
* What is the inner bounding box, meaning the slices for the "rectangular" region where all values are valid?

## Tangent: missing data and dropped frames

So, we also had the topic of missing data some time ago (#256), where we kind of punted out, IIRC with the data set returning zeros for missing portions or similar. Now, with live processing, it will sadly become more common that we drop frames (for example when optimizing a new UDF for live processing and can't keep up, or when the computer used is simply not fast enough to keep up with the chosen computation). With the `valid_nav_mask` or something similar, we now have the chance  to properly communicate this to the UDF in a consistent way. But maybe we actually want to have a separate mask for this, to distinguish dropped frames / missing data from (in the case of STEM) not having visited a given position yet.

A reason for keeping these separately is that UDFs may want to only show results in the "inner bounding box", a rectangular region where all positions have been seen. Otherwise, we may get artifacts/histogram/flickering issues for the live display, at borders etc. If we have a single mask only, and we are missing a single frame, this region will never grow beyond that position.

So maybe, in this PR, `valid_nav_mask` should actually be documented to _include_ positions with missing data, and in the future we can add a second mask especially for the missing data. Agreed?

**Comment by sk1p** [comment]
The `web_api` tests are a bit flaky now, which I managed to reproduce locally, too. I suspect this is related to #1577 and the changes to how the server creates executors etc. - before bugging out, the test was hanging for quite some time in `make_executor`:

```
Thread 1706497 (idle): "LiberTEM-background"
    stop (distributed/utils.py:606)
    close (distributed/client.py:1790)
    __del__ (distributed/client.py:1522)
    _make_invoke_excepthook (threading.py:1324)
    __init__ (threading.py:910)
    __init__ (distributed/utils.py:479)
    _start_unlocked (distributed/utils.py:590)
    start (distributed/utils.py:551)
    start (distributed/client.py:1209)
    __init__ (distributed/client.py:1014)
    connect (libertem/executor/dask.py:701)
    make_executor (libertem/web/state.py:128)
    _do_connect (libertem/web/connect.py:53)
    put (libertem/web/connect.py:45)
    _execute (tornado/web.py:1790)
    _run (asyncio/events.py:80)
    _run_once (asyncio/base_events.py:1914)
    run_forever (asyncio/base_events.py:607)
    run (conftest.py:939)
    _bootstrap_inner (threading.py:1045)
    _bootstrap (threading.py:1002)
```

(where "LiberTEM-background" is the thread used to run the web API in the background for testing - used in fixtures like `base_url`)

**Review by matbryan52 — COMMENTED** [review]
In comments

File: src/libertem/common/buffers.py
Comment (sk1p) [reply src/libertem/common/buffers.py:206]: True, I'll update the type signatures.

File: src/libertem/udf/base.py
Comment (sk1p) [reply src/libertem/udf/base.py:1648]: Agree on the `@staticmethod`, but I'll keep it in `UDF`, as the method needs to appear in the documentation (and as far as I can see, `UDFBase` doesn't, and mostly contains implementation details)

File: src/libertem/udf/base.py
Comment (sk1p) [reply src/libertem/udf/base.py:386]: True, I'll update it to use our `count_nonzero`

File: src/libertem/udf/base.py
Comment (sk1p) [reply src/libertem/udf/base.py:565]: Good catch, I'll also add a test case that ensures `get_valid_nav_mask` always returns `None` in `process_*` (in the case of the inline executor, this apparently wasn't the case, so now I'm explicitly resetting this value to ensure we get a consistent experience across executors)

File: tests/udf/test_valid_mask.py
Comment (sk1p) [reply tests/udf/test_valid_mask.py:43]: Good idea, will do!

**Comment by sk1p** [comment]
@uellue could you have a look specifically at the CoM changes?

File: docs/source/changelog/features/valid-data-masking.rst
Comment (matbryan52) [inline docs/source/changelog/features/valid-data-masking.rst:10]: Rephrasing to help with the release notes later:

```suggestion
* Add endpoints to provide :meth:`~libertem.udf.base.UDF.merge`
  and :meth:`~libertem.udf.base.UDF.get_results` with information
  on which parts of the dataset have already been processed
  (:issue:`1473`). The information is then propagated to the result
  buffers, and UDFs can override the validity mask for results
  using the new :meth:`~libertem.udf.base.UDF.with_mask` method
  (:issue:`1449` :pr:`1593`).
```

File: docs/source/udf/advanced.rst
Comment (matbryan52) [inline docs/source/udf/advanced.rst:343]: Import (unused, but in the test code)


```suggestion
    from libertem.udf import UDF
    from libertem.common.math import count_nonzero    
```

File: docs/source/udf/advanced.rst
Comment (matbryan52) [inline docs/source/udf/advanced.rst:336]:
```suggestion
This is useful if the dataset might contain invalid frames discovered at runtime,
or if intermediate results cannot be fully valid due to missing information (e.g.
calculating gradients across parition boundaries). Anothe example is when using a :code:`kind='single'` buffer which is progressively computed, valid-by-default
might be incorrect if not enough information has been seen (e.g. statistical measures
or a count of values).
```

File: docs/source/udf/advanced.rst
Comment (matbryan52) [inline docs/source/udf/advanced.rst:287]:
```suggestion
In various intermediate states, result buffers may be only partially filled
with valid data. This comes from the fact that we split the dataset into
partitions and then merge results together only as they become available.
When doing live processing we also have the constraint that not all frames
have been acquired yet, and so naturally only positions where we have
received and processed data can be valid in the corresponding result buffer.
```

**Review by matbryan52 — APPROVED** [review]
Looks good to me. If there is anything missing it is a description in the doc of how this differs from the existing `damage` on `UDFResults`, which is linked but a less granular information about validity.

**Comment by sk1p** [comment]
> Looks good to me. If there is anything missing it is a description in the doc of how this differs from the existing `damage` on `UDFResults`, which is linked but a less granular information about validity.

Thanks for having a look, I'll add a short note about the relation to `UDFResults.damage`.

File: src/libertem/common/buffers.py
Comment (uellue) [inline src/libertem/common/buffers.py:207]: The signature says `ndarray` or `bool`, and in the next line we test for `int`, which is a superset of `bool`. Probably the parameters should be documented in the docstring?

* `ndarray` of what types?
* What is the difference between dtype `int` and `bool` ndarrays?

The behavior of NumPy fancy indexing is different with an `int` mask or `bool` mask: With an `int` mask it returns the elements at those indices. I can't quite imagine how that should work with valid data masking, so we should probably restrict the mask to dtype `bool`?

File: src/libertem/udf/com.py
Comment (uellue) [inline src/libertem/udf/com.py:617]: ...see here for a case where a mask of dtype `int` would not work as expected.

File: src/libertem/udf/com.py
Comment (uellue) [inline src/libertem/udf/com.py:694]: Why is this in there, commented out? The variable `mask` is not defined in the previous code, as far as I can tell.

File: tests/udf/test_valid_mask.py
Comment (uellue) [inline tests/udf/test_valid_mask.py:66]: Is there a reason to not use the default fixtures for dataset and inline context?

The pattern repeats in the following tests.

File: tests/udf/test_valid_mask.py
Comment (uellue) [inline tests/udf/test_valid_mask.py:228]: In the same vein we can test that invalid dtypes are rejected

**Review by uellue — CHANGES_REQUESTED** [review]
Looks good, with minor comments on dtypes.

I'll also contribute a test case for correct CoMUDF regression with incomplete input data, so please don't merge yet!

File: src/libertem/udf/com.py
Comment (uellue) [reply src/libertem/udf/com.py]: Went ahead and limited the application of regression to valid data.

File: src/libertem/common/buffers.py
Comment (sk1p) [reply src/libertem/common/buffers.py:207]: True, this was confusing, the type should be `bool` or `np.ndarray`. I'll add documentation to `with_mask`, which is the public interface, to clear this up, and I'll consistently use `bool` instead of `int` (for Python this doesn't really matter, and I somehow got into the habit of using `1` instead of `True` sometimes, but it's clear that there's a difference for `numpy`...). The intention of the interface was to pass in boolean masks, not lists of coordinates, which could be added as a follow-up IMHO.

Comment (uellue) [reply src/libertem/common/buffers.py:207]: Yeah, list of coordinates will also modify the shape and contain duplicates, so I think bool only makes sense for now. :-)

File: tests/udf/test_valid_mask.py
Comment (sk1p) [reply tests/udf/test_valid_mask.py:66]: For one, I've lost overview over the different fixtures we have available. We may need documentation for the most commonly used fixtures in our contributor docs... I used `make_with` as it's just so convenient :) I'll update the tests to use `lt_ctx` instead, which actually does useful things for testing (replaces the plotting with a dummy, and restricts the number of inline threads).

As for the data set, I think it's good to actually have the code in the test, as that means it is self-contained, and you can directly understand the shape of the test data from the test code. When using the fixture, you need to go hunting first. Let me know if you have concerns with this approach, I can easily update the tests to use the `default_raw` fixture instead.

Comment (uellue) [reply tests/udf/test_valid_mask.py:66]: Good point about the data shape! I also really like `make_with()`. :-)

**Review by uellue — APPROVED** [review]
Let's Go, Try Merge!👍

**Pull request merged**
